File: src/main/kotlin/org/skialin/ColorMatrix.kt
package org.skialin

import org.skialin.impl.NativeLoader

/**
 * A 4x5 row-major color matrix, mirroring Skia's `SkColorMatrix`. Feeds directly
 * into [ColorFilter.makeMatrix].
 */
object ColorMatrix {
    fun identity(): FloatArray = ColorMatrixNative.nIdentity()

    fun makeScale(r: Float, g: Float, b: Float, a: Float = 1f): FloatArray = ColorMatrixNative.nScale(r, g, b, a)

    /** `0` desaturates entirely (grayscale), `1` is a no-op. */
    fun makeSaturation(sat: Float): FloatArray = ColorMatrixNative.nSaturation(sat)

    fun postTranslate(mat20: FloatArray, dr: Float, dg: Float, db: Float, da: Float): FloatArray =
        ColorMatrixNative.nPostTranslate(mat20, dr, dg, db, da)

    /** `result = a * b` (`a` is applied after `b`). */
    fun concat(a: FloatArray, b: FloatArray): FloatArray = ColorMatrixNative.nConcat(a, b)
}

private object ColorMatrixNative {
    init {
        NativeLoader.ensureLoaded()
    }

    external fun nIdentity(): FloatArray
    external fun nScale(r: Float, g: Float, b: Float, a: Float): FloatArray
    external fun nSaturation(sat: Float): FloatArray
    external fun nPostTranslate(mat20: FloatArray, dr: Float, dg: Float, db: Float, da: Float): FloatArray
    external fun nConcat(a20: FloatArray, b20: FloatArray): FloatArray
}

File: rust/skialin-core/src/color_matrix.rs
use crate::sys;

/// A 4x5 row-major color matrix, mirroring Skia's `SkColorMatrix`. Feeds
/// directly into [`crate::ColorFilter::matrix`].
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct ColorMatrix(pub [f32; 20]);

impl ColorMatrix {
    pub fn identity() -> Self {
        let mut m = Self([0.0; 20]);
        unsafe { sys::skialin_bridge_ColorMatrix_setIdentity(m.0.as_mut_ptr()) };
        m
    }

    pub fn scale(r_scale: f32, g_scale: f32, b_scale: f32, a_scale: f32) -> Self {
        let mut m = Self([0.0; 20]);
        unsafe { sys::skialin_bridge_ColorMatrix_setScale(m.0.as_mut_ptr(), r_scale, g_scale, b_scale, a_scale) };
        m
    }

    /// Adjusts saturation: `0` desaturates entirely (grayscale), `1` is a no-op.
    pub fn saturation(sat: f32) -> Self {
        let mut m = Self::identity();
        m.set_saturation(sat);
        m
    }

    pub fn set_saturation(&mut self, sat: f32) {
        unsafe { sys::skialin_bridge_ColorMatrix_setSaturation(self.0.as_mut_ptr(), sat) };
    }

    pub fn post_translate(&mut self, dr: f32, dg: f32, db: f32, da: f32) {
        unsafe { sys::skialin_bridge_ColorMatrix_postTranslate(self.0.as_mut_ptr(), dr, dg, db, da) };
    }

    /// `result = a * b` (`a` is applied after `b`).
    pub fn concat(a: &ColorMatrix, b: &ColorMatrix) -> Self {
        let mut out = Self([0.0; 20]);
        unsafe { sys::skialin_bridge_ColorMatrix_setConcat(out.0.as_mut_ptr(), a.0.as_ptr(), b.0.as_ptr()) };
        out
    }
}

impl Default for ColorMatrix {
    fn default() -> Self {
        Self::identity()
    }
}
